### PR TITLE
Car docs bot: diff on generated readme file

### LINKED ISSUE #32690: Car docs bot: diff on generated readme file
We have a bot that uses the car docs Python data structures and prints a pretty diff to each PR, if it changes any car docs. That could be vehicle name, supported trim, minimum steering speed, parts, or detail sentence for the website.

Current bot output: https://github.com/commaai/openpilot/pull/31888#issuecomment-2000768679

Our current solution requires checkouts and pickling, which is fragile: [print_docs_diff.py](https://github.com/commaai/openpilot/blob/master/selfdrive/debug/print_docs_diff.py) and [selfdrive_tests.yaml#L256-L308](https://github.com/commaai/openpilot/blob/c514ee7212f7ef43c7451d6e638dd54ec0643dcc/.github/workflows/selfdrive_tests.yaml#L256-L308).

Solution
---

Just use the base and head commit CARS.md files in the PR, for example:

```
wget -O master_table.md https://raw.githubusercontent.com/commaai/openpilot/master/docs/CARS.md
diff master_table.md docs/CARS.md
```

See the generated CARS.md README file: https://github.com/commaai/openpilot/blob/master/docs/CARS.md

Diffing behavior:
- [ ] Should be agnostic to the columns and other changes (you can add or remove without issues)
- [ ] Minimal understanding of the file, which makes this small, simple, and robust. You may need special handling of the detail sentence however, which is fine.
- [ ] Changes to column values should be picked up as a column change, not addition and removal. GitHub already does that for us!

### PR DESCRIPTION
**Pull request by marcellofuschi**
Closes https://github.com/commaai/openpilot/issues/32690

- Doesn't use pickle, it just compares the markdown files (the one in master and the one in a PR).
- Only the car docs table is compared. The rest of the file is ignored.
- It relies on the (Make, Model) pair from the table to identify a row. If a column value was changed but the pair remained the same, that's shown as a column change in the PR comment (as opposed to an addition+removal).
- Adds the Detail sentence as an expandable button in the docs (like "Hardware Needed").

Here's an example result: https://github.com/marcellofuschi/openpilot/pull/2#issuecomment-2284364319
Detail sentence isn't visible in the example because that column is not in master yet.

### CODE REVIEW

**Comment by marcellofuschi** [comment]
Need to fix: the code for parsing the markdown table is too naive and can't handle escaped pipes correctly. Will do it in a bit.

**Comment by sshane** [comment]
The sentence is not very easily to read, is there something we can do about that? Tooltip, comment? We don't show it now, so it's not a requirement to show it now.

image

And does this satisfy this point in the issue?
>Changes to column values should be picked up as a column change, not addition and removal. GitHub already does that for us!

**Comment by marcellofuschi** [comment]
>The sentence is not very easy to read, is there something we can do about that?

Pushed a change that should improve that. I automatically add some `&nbsp;` to the column value so that once rendered it prefers overflowing with scroll (up to a certain point).

>And does this satisfy this point in the issue?
>"Changes to column values should be picked up as a column change, not addition and removal."

I think so, as long as the changes aren't made to the columns I use for identifying a row (i.e., `Make` and `Model`).

---

I also added a length limit to the generated comment, so that it won't show more than 10 rows per category, with an indication of the number of hidden rows. I believe the comment would get truncated anyway if too long so it's better to handle it more gracefully.

---

>Need to fix: the code for parsing the markdown table is too naive and doesn't handle potential escaped pipes.

Still gotta fix this.

**Comment by sshane** [comment]
Getting:

```python
batman@workstation-shane:~/openpilot/selfdrive/debug$ python3 print_docs_diff.py 
  File "/home/batman/openpilot/selfdrive/debug/print_docs_diff.py", line 22
    results = {f'{r['Make']} {r['Model']}': r for r in results}
                     ^^^^
SyntaxError: f-string: unmatched '['
```

File: selfdrive/debug/print_docs_diff.py [hunk 1]
```diff
@@ -1,120 +1,93 @@
...
-def build_column_diff(base_car, new_car):
-  row_builder = []
-  for column in Column:
-    base_column = base_car.get_column(column, STAR_ICON, VIDEO_ICON, FOOTNOTE_TAG)
-    new_column = new_car.get_column(column, STAR_ICON, VIDEO_ICON, FOOTNOTE_TAG)
+  match = re.search(r"(\d+)\s+supported\s+cars\s*\n([\s\S]*?\|.*?\|[\s\S]*?\|[-:]+\|[\s\S]*?\n(?:[^\S\r\n]*\S.*\n)*\n?)", docs_content, re.IGNORECASE)
+  if not match:
+    raise RuntimeError("Couldn't find the car docs Markdown table.")
+  num_cars = int(match.group(1))
+  md_table = match.group(2).strip()
```
Comment (sshane) [inline selfdrive/debug/print_docs_diff.py:20]:
```suggestion
  num_cars, md_table = match.groups()
```

I like unpacking in one line for the best readability, and you know immediately where the usage of `match` ends.

File: selfdrive/debug/print_docs_diff.py [hunk 2]
```diff
@@ -1,120 +1,93 @@
...
+  with open(new_docs_path) as new_docs_file:
+    new_docs_content = new_docs_file.read()
 
-    for new_car, base_car in car_changes:
-      # Column changes
-      row_diff = build_column_diff(base_car, new_car)
-      if ARROW_SYMBOL in row_diff:
-        changes["column"].append(row_diff)
+  base_headers, base_cars = get_cars_docs_in_markdown(base_docs_content)
+  new_headers, new_cars = get_cars_docs_in_markdown(new_docs_content)
```
Comment (sshane) [inline selfdrive/debug/print_docs_diff.py:63]: It makes more sense if `new_docs_content` is hardcoded using BASEDIR + CARS.md file locally, and `base_docs_content` should be passed in as a path, using the base commit on the PR (downloaded in the Action)

File: selfdrive/debug/print_docs_diff.py [hunk 3]
```diff
@@ -1,120 +1,93 @@
...
   # Print diff
   if any(len(c) for c in changes.values()):
     markdown_builder = ["### ⚠️ This PR makes changes to [CARS.md](../blob/master/docs/CARS.md) ⚠️"]
 
-    for title, category in (("## 🔀 Column Changes", "column"), ("## ❌ Removed", "removals"),
-                            ("## ➕ Added", "additions"), ("## 📖 Detail Sentence Changes", "detail")):
-      if len(changes[category]):
+    for title, category in (("## 🔀 Column Changes", "column"), ("## ❌ Removed", "removals"), ("## ➕ Added", "additions")):
+      ordered_headers = base_headers if category != 'additions' else new_headers
+      max_rows_to_show = 10
```
Comment (sshane) [inline selfdrive/debug/print_docs_diff.py:77]: Unless there is a problem specific to this PR, let's keep this printing behavior the same (no row limit).

**Comment by marcellofuschi** [comment]
> I would say it's okay to hide the sentence for now, it changes the style of the table too much and pushes the other columns near the end off the screen. Another user put it in the markdown file as comments, with some small special handling.

This was partly because I had also enlarged the `Supported Package` column. I reverted it now, so the [table](https://github.com/marcellofuschi/openpilot/blob/car-docs-diff/docs/CARS.md) should look a bit more familiar. Let me know if this could do it.

Regarding `re.sub`s, the easier to remove would be the first one, though I think it's more reliable to also check for word boundary before the appearance of `assets/` when you replace. Let me know if you agree, otherwise I'll change it to `.replace('assets/', 'https://media.githubusercontent.com/media/commaai/openpilot/master/docs/assets/')`

I'll look into the other points you raised (thanks for the feedback).

**Comment by marcellofuschi** [comment]
>Getting this, why doesn't CI catch it?:

I reproduced the error, I think it has to do with quotes used wrong. Demo example:

Screenshot 2024-08-13 at 4 04 30 PM


Fixed it in the code, so it should be okay now. Don't know why CI wasn't catching it.

File: selfdrive/debug/print_docs_diff.py (hunk 2 above)
Comment (marcellofuschi) [reply selfdrive/debug/print_docs_diff.py:64]: I made the change (https://github.com/commaai/openpilot/pull/33273/commits/ce8c77cdb15874ca3eea2555e4830202a58cfa00) and tested it in my fork. Is this what you had in mind?

File: .github/workflows/selfdrive_tests.yaml [hunk 4]
```diff
@@ -289,9 +289,9 @@ jobs:
           ref: ${{ github.event.pull_request.base.ref }}
```
Comment (sshane) [inline .github/workflows/selfdrive_tests.yaml:289]: We want to get rid of the two checkouts in this PR and just use curl/wget

Comment (marcellofuschi) [reply .github/workflows/selfdrive_tests.yaml]: Done (https://github.com/commaai/openpilot/pull/33273/commits/ddbe96471e627abb5fe22f0e8df54cd1a9e25260)

**Comment by marcellofuschi** [comment]
Rebased after the recent changes, and tested again [here](https://github.com/marcellofuschi/openpilot/pull/6)

The Detail sentence is missing from the docs because `docs_definitions.py` was moved to `opendbc`, so the diff doesn't currently involve Detail sentence.

I think it looked fine when we added it to the table as a new column (with formatting done through `&nbsp;` usage). I opened a [PR](https://github.com/commaai/opendbc/pull/1182) to `opendbc` to re-add it.

**Pull request closed without merging**